apidef/rate_limit_smoothing.go

```diff
@@ -0,0 +1,102 @@
+package apidef
+
+import (
+	"errors"
+	"fmt"
+	"time"
+)
+
+// RateLimitSmoothing holds the rate smoothing configuration in effect.
+type RateLimitSmoothing struct {
+	// Enabled if true will enable rate limit smoothing.
+	Enabled bool `json:"enabled" bson:"enabled"`
+
+	// Treshold is the value above which gateway will apply smoothing.
+	Threshold int64 `json:"threshold" bson:"threshold"`
+
+	// Triger holds a value between 0..1 and is used as a percentage value of the
+	// rate limit step to reach, before triggering a SmoothingUp event. Similarly the
+	// value is also used to decrease the allowance when rate limits decrease.
+	//
+	// Example:
+	//
+	// - Allowance 600,
+	// - Current rate 500,
+	// - Step 100,
+	// - Trigger 0.5
+	//
+	// To trigger a RateLimitSmoothingUp event, the rate needs to exceed 550.
+	// The new allowance is calculated as: `Allowance - Trigger * Step`.
+	//
+	// To trigger a RateLimitSmoothingDown event, rate needs to fall below 450.
+	// The new allowance is calculated as: `Allowance - Trigger - (Trigger*Step)`
```

Comment:
The threshold to trigger the smoothing down event is current rate (500) - (trigger(0.5) * step(100)) = 450, yes?
Additionally, the new allowance would then be calculated as 600 - 0.5 - (0.5*100) = 549.5, yes?

So the allowance always decreases. When rate limit decreases the allowance decreases. However, if the rate limit increases again within the boundary of the originally set allowance the allowance never increases, correct?